src/main/haskell/kore/src/Kore/Step/Simplification/And.hs

```diff
@@ -0,0 +1,238 @@
+        , SimplificationProof
+        )
+
+{-|'makeEvaluate' simplifies an 'And' of 'ExpandedPattern's.
+
+See the comment for 'simplify' to find more details.
+-}
+makeEvaluate
+    ::  ( MetaOrObject level
+        , SortedVariable variable
+        , Show (variable level)
+        , Ord (variable level)
+        , Ord (variable Meta)
+        , Ord (variable Object)
+        , IntVariable variable
+        , Hashable variable
+        )
+    => MetadataTools level StepperAttributes
+    -> ExpandedPattern level variable
+    -> ExpandedPattern level variable
+    -> IntCounter (ExpandedPattern level variable, SimplificationProof level)
+makeEvaluate
+    tools first second
+  | ExpandedPattern.isBottom first || ExpandedPattern.isBottom second =
+    return (ExpandedPattern.bottom, SimplificationProof)
+  | ExpandedPattern.isTop first =
+    return (second, SimplificationProof)
+  | ExpandedPattern.isTop second =
+    return (first, SimplificationProof)
+  | otherwise =
+    makeEvaluateNonBool tools first second
+
+makeEvaluateNonBool
+    ::  ( MetaOrObject level
+        , SortedVariable variable
+        , Show (variable level)
+        , Ord (variable level)
+        , Ord (variable Meta)
+        , Ord (variable Object)
+        , IntVariable variable
+        , Hashable variable
+        )
+    => MetadataTools level StepperAttributes
+    -> ExpandedPattern level variable
+    -> ExpandedPattern level variable
+    -> IntCounter (ExpandedPattern level variable, SimplificationProof level)
+makeEvaluateNonBool
+    tools
+    ExpandedPattern
+        { term = firstTerm
+        , predicate = firstPredicate
+        , substitution = firstSubstitution
+        }
+    ExpandedPattern
+        { term = secondTerm
+        , predicate = secondPredicate
+        , substitution = secondSubstitution
+        }
+  = do -- IntCounter monad
+    (   PredicateSubstitution
+            { predicate = mergedPredicate
+            , substitution = mergedSubstitution
+            }
+        , _proof
+        ) <- mergePredicatesAndSubstitutions
+            tools
+            [firstPredicate, secondPredicate]
+            [firstSubstitution, secondSubstitution]
+    return
+        ( ExpandedPattern
+            { term = give sortTools $ makeTermAnd firstTerm secondTerm
+            , predicate = mergedPredicate
+            , substitution = mergedSubstitution
+            }
+        , SimplificationProof
+        )
+  where
+    sortTools = MetadataTools.sortTools tools
+
+makeTermAnd
+    ::  ( MetaOrObject level
+        , SortedVariable variable
+        , Given (SortTools level)
+        , Show (variable level)
+        , Ord (variable level)
+        )
+    => PureMLPattern level variable
+    -> PureMLPattern level variable
+    -> PureMLPattern level variable
+makeTermAnd b@(Bottom_ _) _ = b
+makeTermAnd (Top_ _) term = term
+-- TODO: (partial) unification / other simplifications
+makeTermAnd first second = makeTermAndSecond first second
```

Comment:
Why not in the same function?

Reply:
> Because the Haskell pattern matching stuff is o(n^2) and it will emit a warning of it taking over 2M iterations to check a function that uses the underscore matchers (e.g. Bottom_) on at least two arguments.

The exhaustiveness checker does not work well for recursive matching in general, but the situation does get much worse with pattern synonyms. We can write the `case`s ourselves, which is still reasonably simple:
```.hs
makeTermAnd first second =
    case first of
        Bottom_ _ -> first
        Top_ _ -> second
        _ -> case second of
            Bottom_ _ -> second
            Top_ _ -> first
            _ -> mkAnd first second
```

(I think pattern matching on the LHS of `=` was a huge design mistake for Haskell, and recursive matching in general, but that's a long rant.) Pattern synonyms are supported better in GHC 8.4, but we don't have that on Jenkins at the moment.